tools/xcarchivetool/xcarchivetool.bzl

```
@@ -0,0 +1,89 @@
+"""
+A rule for re-packaging a .ipa into a .xcarchive.
+"""
+
+load(
+    "@build_bazel_rules_apple//apple:providers.bzl",
+    "AppleBundleInfo",
+    "AppleDsymBundleInfo",
+)
+
+def _xcarchive_impl(ctx):
+    """
+    Implementation for xcarchive.
+
+    This rule uses the providers from the app target to re-package an .ipa or .app bundle into a
+    .xcarchive. The .xcarchive is a directory that contains the app bundle, dSYM and other metadata.
+
+    It calls out to a binary to do the actual re-packaging.
+    """
+    bundle_info = ctx.attr.app[AppleBundleInfo]
+    dsym_info = ctx.attr.app[AppleDsymBundleInfo]
+    xcarchive = ctx.actions.declare_directory("%s.xcarchive" % bundle_info.bundle_name)
+
+    arguments = ctx.actions.args()
+    arguments.add("--info_plist", bundle_info.infoplist.path)
+    arguments.add("--archive", bundle_info.archive.path)
+    arguments.add("--output", xcarchive.path)
+
+    for dsym in dsym_info.direct_dsyms:
+        arguments.add("--dsym", dsym.path)
+
+    ctx.actions.run(
+        inputs = depset([
+            bundle_info.archive,
+            bundle_info.infoplist,
+        ] + dsym_info.direct_dsyms),
+        outputs = [xcarchive],
+        executable = ctx.executable._make_xcarchive,
+        arguments = [arguments],
+        mnemonic = "XCArchive",
+    )
+
+    return [
+        DefaultInfo(files = depset([xcarchive])),
+    ]
+
+xcarchive = rule(
+    implementation = _xcarchive_impl,
+    attrs = {
+        "app": attr.label(
+            providers = [
+                AppleBundleInfo,
+                AppleDsymBundleInfo,
+            ],
+            doc = """
+            The label to a target to re-package into a .xcarchive. For example, an `ios_application` target.
+            """,
+        ),
+        "_make_xcarchive": attr.label(
+            default = Label("@build_bazel_rules_apple//tools/xcarchivetool:make_xcarchive"),
+            executable = True,
+            cfg = "exec",
+            doc = """
+            An executable binary that can re-package a bundle into a .xcarchive.
+            """,
+        ),
+    },
+    doc = """
+    Re-packages an .ipa or .app bundle into a .xcarchive.
+    This rule uses the providers from the app target to construct the required metadata for the .xcarchive.
+
+    Example:
+
+    ````python
+    load("@build_bazel_rules_apple//apple:xcarchive.bzl", "xcarchive")
+
+    ios_application(
+        name = "App",
+        bundle_id = "com.example.my.app",
+        ...
+    )
+
+    xcarchive(
+        name = "App.xcarchive",
+        app = ":App",
+    )
+    ````
+    """,
```

```suggestion
    doc = """\
Re-packages an .ipa or .app bundle into a .xcarchive.

This rule uses the providers from the app target to construct the required
metadata for the .xcarchive.

Example:

```starlark
load("@build_bazel_rules_apple//apple:xcarchive.bzl", "xcarchive")

ios_application(
    name = "App",
    bundle_id = "com.example.my.app",
    ...
)

xcarchive(
    name = "App.xcarchive",
    app = ":App",
)
```
""",

---

(I broke the GiHub suggestion, but please follow the doc formatting from other rules)